adapter/prometheus/prometheus_test.go

```diff
@@ -0,0 +1,294 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+
+	"istio.io/mixer/adapter/prometheus/config"
+	"istio.io/mixer/pkg/adapter"
+	"istio.io/mixer/pkg/adapter/test"
+)
+
+type testServer struct {
+	server
+
+	errOnStart bool
+}
+
+func (t testServer) Start(adapter.Logger) error {
+	if t.errOnStart {
+		return errors.New("could not start server")
+	}
+	return nil
+}
+
+var (
+	gaugeNoLabels = adapter.MetricDefinition{
+		Name:        "/funky::gauge",
+		Description: "funky all the time",
+		Kind:        adapter.Gauge,
+		Labels:      map[string]adapter.LabelKind{},
+	}
+
+	counterNoLabels = adapter.MetricDefinition{
+		Name:        "the.counter",
+		Description: "count all the tests",
+		Kind:        adapter.Counter,
+		Labels:      map[string]adapter.LabelKind{},
+	}
+
+	counter = adapter.MetricDefinition{
+		Name:        "special_counter",
+		Description: "count all the special tests",
+		Kind:        adapter.Counter,
+		Labels: map[string]adapter.LabelKind{
+			"bool":   adapter.Bool,
+			"string": adapter.String,
+			"email":  adapter.EmailAddress,
+		},
+	}
+
+	unknown = adapter.MetricDefinition{
+		Name:        "unknown",
+		Description: "unknown",
+		Kind:        adapter.Gauge - 2,
+		Labels:      map[string]adapter.LabelKind{},
+	}
+
+	counterVal = adapter.Value{
+		Name: counter.Name,
+		Labels: map[string]interface{}{
+			"bool":   true,
+			"string": "testing",
+			"email":  "test@istio.io",
+		},
+		Kind:        adapter.Counter,
+		MetricValue: float64(45),
+	}
+
+	gaugeVal = newGaugeVal(int64(993))
+)
+
+func TestInvariants(t *testing.T) {
+	test.AdapterInvariants(Register, t)
+}
+
+func TestFactory_NewMetricsAspect(t *testing.T) {
+	f := newFactory(&testServer{})
+
+	tests := []struct {
+		name    string
+		metrics []adapter.MetricDefinition
+	}{
+		{"No Metrics", []adapter.MetricDefinition{}},
+		{"One Gauge", []adapter.MetricDefinition{gaugeNoLabels}},
+		{"One Counter", []adapter.MetricDefinition{counterNoLabels}},
+		{"Multiple Metrics", []adapter.MetricDefinition{counterNoLabels, gaugeNoLabels}},
+		{"With Labels", []adapter.MetricDefinition{counter}},
+	}
+
+	for _, v := range tests {
+		t.Run(v.name, func(t *testing.T) {
+			if _, err := f.NewMetricsAspect(test.NewEnv(t), &config.Params{}, v.metrics); err != nil {
+				t.Errorf("NewMetricsAspect() => unexpected error: %v", err)
```

Comment:
Small nit: I like @mandarjog's style of including the test index in the error so I know immediately which case failed. Can you add that to the error strings here and elsewhere? Atm there's no way for me to know which of the cases is broken without manually debugging.